File: solutions/ocaml/085-find-k-closest-elements.ml
(*
  Problem 85: Find K Closest Elements (LeetCode 658)
  Difficulty: Med
  Language: OCaml
*)
let find_closest_elements arr k x =
  let a = Array.of_list arr in
  let rec go lo hi =
    if hi - lo = k then Array.to_list (Array.sub a lo k)
    else if abs (a.(lo) - x) <= abs (a.(hi-1) - x) then go lo (hi-1)
    else go (lo+1) hi
  in go 0 (Array.length a)

File: solutions/rust/085-find-k-closest-elements.rs
/**
 * Problem 85: Find K Closest Elements (LeetCode 658)
 * Difficulty: Med
 * Language: Rust
 */
fn find_closest_elements(arr: &[i32], k: usize, x: i32) -> Vec<i32> {
    let (mut lo, mut hi) = (0usize, arr.len());
    while hi - lo > k {
        if (arr[lo] - x).abs() <= (arr[hi - 1] - x).abs() { hi -= 1; }
        else { lo += 1; }
    }
    arr[lo..hi].to_vec()
}
